presto-main/src/main/java/com/facebook/presto/failureDetector/HeartbeatFailureDetector.java

```diff
@@ -143,6 +148,31 @@ public void shutdown()
                 .collect(toImmutableSet());
     }
 
+    @Override
+    public State getState(HostAddress hostAddress)
+    {
+        for (MonitoringTask task : tasks.values()) {
+            if (hostAddress.equals(fromUri(task.uri))) {
+                if (!task.isFailed()) {
+                    return State.ALIVE;
+                }
+                else {
+                    Exception lastException = task.getStats().getLastFailureException();
+                    if (lastException instanceof SocketTimeoutException || lastException instanceof UnknownHostException) {
+                        return State.GONE;
+                    }
+                    else if (lastException instanceof ConnectException) {
+                        return State.UNRESPONSIVE;
+                    }
+                    else {
+                        return State.UNKNOWN;
+                    }
+                }
+            }
+        }
+        return State.UNKNOWN;
```

Comment:
@raghavsethi  `return` vs. `if-else` is an interesting question.  Looks like nowadays pepole are preferring using early return :): https://www.quora.com/Which-one-is-better-using-if-else-or-return